test/tests/functional/pbs_soft_walltime.py

```diff
@@ -687,3 +709,336 @@ def test_preempt_order(self):
+        """
+        Test that soft walltime is set by runjob hook
+        """
+
+        hook_body = \
+            """import pbs
+e = pbs.event()
+e.job.Resource_List["soft_walltime"] = 5
+e.accept()
+"""
+        a = {'event': 'runjob', 'enabled': 'True'}
+        self.server.create_import_hook("que", a, hook_body)
+
+        J = Job(TEST_USER)
+        jid = self.server.submit(J)
+
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 5}, id=jid)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+    def test_soft_modifyjob_hook(self):
+        """
+        Test that soft walltime is set by modifyjob hook
+        """
+
+        hook_body = \
+            """import pbs
+e = pbs.event()
+e.job.Resource_List["soft_walltime"] = 15
+e.accept()
+"""
+        a = {'event': 'modifyjob', 'enabled': 'True'}
+        self.server.create_import_hook("que", a, hook_body)
+
+        J = Job(TEST_USER)
+        jid = self.server.submit(J)
+
+        self.server.expect(
+            JOB, 'Resource_List.soft_walltime', op=UNSET, id=jid)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        self.server.alterjob(jid, {'Resource_List.soft_walltime': 5})
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 15}, id=jid)
+
+    def test_walltime_default(self):
+        """
+        Test soft walltime behavior with hard walltime is same
+        even if set under resource_default
+        """
+
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'Resources_default.soft_walltime': '15'})
+
+        J = Job(TEST_USER, attrs={'Resource_List.walltime': 15})
+        jid = self.server.submit(J)
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 15}, id=jid)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        self.server.deljob(jid, wait=True)
+
+        J = Job(TEST_USER, attrs={'Resource_List.walltime': 16})
+        jid1 = self.server.submit(J)
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 15}, id=jid1)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        self.server.deljob(jid1, wait=True)
+
+        try:
+            J = Job(TEST_USER, attrs={'Resource_List.walltime': 10})
+            jid1 = self.server.submit(J)
+        except PtlSubmitError as e:
+            self.assertTrue("illegal resource" in e.msg[0])
+
+    def test_soft_suspension(self):
+        """
+        Test that if job is suspended soft_walltime will not get extended
+        """
+
+        J = Job(TEST_USER, {'Resource_List.walltime': 300})
+        jid = self.server.submit(J)
+        self.server.alterjob(jid, {'Resource_List.soft_walltime': 7})
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        self.server.sigjob(jid, signal='suspend')
+        self.server.expect(JOB, {'job_state': 'S'}, id=jid)
+
+        time.sleep(7)
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 7}, id=jid)
```

Comment:
I updated to estimated.soft_walltime but it is not set till it is extended. is this expected?

Reply:
Correct, it is not set until it is extended.  Before that, the estimate is the Resource_List.soft_walltime.  Come to think of it, this could cause some confusion in the case of a requeued job.  If a job has an estimated.soft_walltime set and is requeued, it will remain set.  When the job is run again, it won't be reset until the Resource_List.soft_walltime is exceeded.  Of course this plagues all the estimated resources.  If a job has an estimated.start_time and is requeued, it remains set.